KTOR-2210 Make base name of micrometer metrics configurable

asevich opened the pull request:
**Subsystem**
Server, Features/Micrometer

**Motivation**
When Ktor service is running in the environment with services written in other frameworks like Spring Boot, having standardized metric names is important for monitoring. Static, non-configurable "ktor.http.server" base name is inconvenient and breaks existing monitoring setup or at least requires special handling for Ktor services.

**Solution**
Make it configurable

asevich commented:
@cy6erGn0m could you please have a look and merge or reject it?

rsinukov reviewed (changes requested):
Hi and thanks for your contribution.
Can you please rebase on `1.6.0-eap` branch, run `./gradlew apiDump`, and commit new API files?
Otherwise, LGTM.

asevich commented:
well, I did my best, but I'm sometimes get confused with rebase. If that doesn't look good then I will just close this PR and open new one with 1.6.0-eap as base

rsinukov commented:
@asevich you need to do interactive rebase and drop all the commits that are not related. 
`git rebase -i origin/1.6.0-eap` and replace `pick` with `drop` (or just `d`) in every commit, except the one that you need.

asevich commented:
@rsinukov thank you for the tip! 
I think now it should be fine

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/test/io/ktor/metrics/micrometer/MicrometerMetricsTests.kt:
```diff
@@ -323,6 +323,30 @@ class MicrometerMetricsTests {
+        assertEquals("$newBaseName.requests.active", MicrometerMetrics.activeGaugeName)
+    }
+
+
```
checkstyle complains about this extra blank line

rsinukov commented on ktor-features/ktor-metrics-micrometer/api/ktor-metrics-micrometer.api:
```diff
@@ -1,18 +1,18 @@
 public final class io/ktor/metrics/micrometer/MicrometerMetrics {
 	public static final field Feature Lio/ktor/metrics/micrometer/MicrometerMetrics$Feature;
-	public static final field activeGaugeName Ljava/lang/String;
```
This change is a problem. We can't have any removals in non-major versions. You need to keep old constants, but deprecate them and have some other meaningful name for new properties.

rsinukov reviewed (changes requested):
Thanks! Just a couple of things.

rsinukov commented:
Can you rebase one more time, please? We rebased `1.6.0-eap` on `main` today, so no your PR has some extra commits again.

asevich commented:
I've reverted changes to existing api and added new properties, but naming is a bit pain in the back. Maybe `configurableActiveGaugeName` and `configurableRequestTimerName` would be better? I'm lost here
Other thing is - I've added in every test an assertion to verify, that same gauge and timer are accessible via old and new properties, iff metricsBaseName isn't changed. I hope that was a correct thing to do

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/test/io/ktor/metrics/micrometer/MicrometerMetricsTests.kt:
```diff
@@ -323,6 +331,29 @@ class MicrometerMetricsTests {
 
+    @Test
+    fun `Throws exception when base name is not defined`(): Unit = withTestApplication {
+        val exception = assertFails {
```
`assertFailsWith<IllegalArgumentException> { ...`

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/test/io/ktor/metrics/micrometer/MicrometerMetricsTests.kt:
```diff
@@ -259,15 +266,16 @@ class MicrometerMetricsTests {
             assertNull(throwableCaughtInEngine)
-            assertTrue(noHandlerHandledReqeust)
+            assertTrue(noHandlerHandledRequest)
+            testRegistry.assertEqualRequestTimers()
```
Not sure if we need it for every test. I think it's better to add an extra test that verifies new properties have an old value by default and that's it.

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt:
```diff
@@ -166,25 +169,52 @@ public class MicrometerMetrics private constructor(
          * Active requests gauge name
          */
+        @Deprecated(
+            "static gauge name deprecated",
```
`is deprecated`

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt:
```diff
@@ -166,25 +169,52 @@ public class MicrometerMetrics private constructor(
          * Request time timer name
          */
+        @Deprecated(
+            "static request time timer name deprecated",
```
`is deprecated`

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt:
```diff
@@ -166,25 +169,52 @@ public class MicrometerMetrics private constructor(
      * Micrometer feature installation object
      */
     public companion object Feature : ApplicationFeature<Application, Configuration, MicrometerMetrics> {
+        @Deprecated("static metrics base name deprecated", level = DeprecationLevel.WARNING)
```
no need to deprecate it. it's private, no one will use it anyway. But maybe rename it to `defaultBaseName`?

rsinukov reviewed (approved):
LGTM, just some very minor things

rsinukov commented on ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt:
```diff
@@ -88,6 +88,8 @@ public class MicrometerMetrics private constructor(
      * */
     public class Configuration {
 
+        public var baseName: String = Feature.defaultBaseName
```
the very last thing: can you add kdoc to this and other public properties, please?

asevich commented:
I'm not really sure where to add the kdoc, so I just did monkey copy/paste job. Please tell me, if you meant something different

rsinukov commented:
Thanks! Merged!
